drivers/all_neighbors_system_dashboard/app/models/all_neighbors_system_dashboard/enrollment_attribute_calculations.rb

```diff
@@ -0,0 +1,167 @@
+        12 => 444, # Referral to Joint TH-RRH project/unit/resource opening
+        13 => 445, # Referral to RRH project resource opening
+        14 => 446, # Referral to PSH project resource opening
+        15 => 447, # Referral to Other PH project/unit/resource opening
+        # 16 => 'Referral to emergency assistance/flex fund/furniture assistance',
+        17 => 1114, # Referral to Emergency Housing Voucher (EHV)
+        # 18 => # Referral to a Housing Stability Voucher
+      }.freeze
+
+      def household_type(enrollment)
+        return 'Children Only' if enrollment.children_only?
+        return 'Adults and Children' if enrollment.other_clients_under_18.positive? && enrollment.age > 17
+
+        'Adults Only'
+      end
+
+      def prior_living_situation_category(enrollment)
+        case enrollment.living_situation
+        when 30, 17, 37, 8, 9, 99
+          'Unknown'
+        when 16
+          'Unsheltered'
+        when 1, 18, 2
+          'Sheltered'
+        when 15, 6, 7, 25, 4, 5
+          'Institutional'
+        when 29, 14, 32, 13, 27, 12, 22, 35, 36, 23, 26, 28, 19, 3, 31, 33, 34, 10, 20, 21, 11
+          'Housed'
+        when 24
+          'Deceased'
+        else
+          'ERROR'
+        end
+      end
+
+      def exit_date(filter, enrollment)
+        date = enrollment.exit_date
+        return nil if date.present? && date > filter.end_date
+
+        date
+      end
+
+      def adjusted_edit_date(filter, enrollment)
+        date = enrollment.exit_date
+        return filter.end_date if date.blank?
+        return nil if date.present? && date > filter.end_date
+
+        date
+      end
+
+      def exit_type(enrollment)
+        case enrollment.destination
+        when nil
+          nil
+        when 26, 11, 21, 3, 10, 28, 20, 19, 22, 23, 31, 33, 34
+          'Permanent'
+        when 6, 24, 15
+          'Excludable'
+        when 8, 9, 99, 30, 17
+          'Unknown'
+        else
+          'Non-Permanent'
+        end
+      end
+
+      def relationship(enrollment)
+        case enrollment.relationship_to_ho_h
+        when 1
+          'SL'
+        when 2
+          'DC'
+        when 3
+          'SP'
+        when 4
+          'F'
+        when 5
+          'O'
+        end
+      end
+
+      def gender(enrollment)
+        HudUtility.gender(enrollment.client.gender_binary)
+      end
+
+      def primary_race(enrollment)
+        HudUtility.race(enrollment.client.pit_race, multi_racial: true)
+      end
+
+      def race_list(enrollment)
+        enrollment.client.race_description
+      end
+
+      def ce_info(filter, housing_enrollment)
+        household_enrollments = GrdaWarehouse::ServiceHistoryEnrollment.
+          entry.
+          where(household_id: housing_enrollment.household_id, data_source_id: housing_enrollment.data_source_id).
+          where(she_t[:exit_date].eq(nil).or(she_t[:exit_date].gteq(filter.start_date)))
```

Comment:
might we want to batch these up?  Seems like this is a potentially massive N+1

Reply:
You're right -- this, and `return_date` are both N+1s, that I need to think about... I didn't really think about performance at all, I was just tracking the existing SQL -- I'm betting some pre-loading is probably also useful.